internal/test/cleanup/cleanup.go

```diff
@@ -189,3 +195,73 @@ func NutanixTestResources(clusterName, endpoint, port string, insecure, ignoreEr
 	}
 	return nil
 }
+
+// TinkerbellTestResources cleans up machines by powering them down.
+func TinkerbellTestResources(ctx context.Context, inventoryCSVFilePath string, ignoreErrors bool) error {
+	hardwarePool, err := api.NewHardwareMapFromFile(inventoryCSVFilePath)
+	if err != nil {
+		return fmt.Errorf("failed to create hardware map from inventory csv: %v", err)
+	}
+
+	logger.Info("Powering off hardware: %+v", hardwarePool)
+	return powerOffHardwarePool(ctx, hardwarePool, ignoreErrors)
+}
+
+func powerOffHardwarePool(ctx context.Context, hardware map[string]*hardware.Machine, ignoreErrors bool) error {
+	errList := []error{}
+	for _, h := range hardware {
+		if err := powerOffHardware(ctx, h, ignoreErrors); err != nil {
+			errList = append(errList, err)
+		}
+	}
+
+	if len(errList) > 0 {
+		return fmt.Errorf("failed to power off %d hardware: %+v", len(errList), errors.NewAggregate(errList))
+	}
+
+	return nil
+}
+
+func powerOffHardware(ctx context.Context, h *hardware.Machine, ignoreErrors bool) (reterror error) {
+	bmcClient := test.NewBmclibClient(logr.Discard(), h.BMCIPAddress, h.BMCUsername, h.BMCPassword)
+
+	if err := bmcClient.Open(ctx); err != nil {
+		md := bmcClient.GetMetadata()
+		logger.Info("Warning: Failed to open connection to BMC: %v, hardware: %v, providersAttempted: %v, failedProviderDetail: %v", err, h.BMCIPAddress, md.ProvidersAttempted, md.SuccessfulOpenConns)
+		return handlePowerOffHardwareError(err, ignoreErrors)
+	}
+
+	md := bmcClient.GetMetadata()
+	logger.Info("Connected to BMC: hardware: %v, providersAttempted: %v, successfulProvider: %v", h.BMCIPAddress, md.ProvidersAttempted, md.SuccessfulOpenConns)
+
+	defer func() {
+		if err := bmcClient.Close(ctx); err != nil {
+			md := bmcClient.GetMetadata()
+			logger.Info("Warning: BMC close connection failed: %v, hardware: %v, providersAttempted: %v, failedProviderDetail: %v", err, h.BMCIPAddress, md.ProvidersAttempted, md.FailedProviderDetail)
+			reterror = handlePowerOffHardwareError(err, ignoreErrors)
+		}
+	}()
+
+	state, err := bmcClient.GetPowerState(ctx)
+	if err != nil {
+		state = "unknown"
+	}
+	if strings.Contains(strings.ToLower(state), "off") {
+		return nil
+	}
+
+	if _, err := bmcClient.SetPowerState(ctx, "off"); err != nil {
```

Comment:
We talked internally about the importance of machines to be off (because of duplicate DHCP servers running) before each test run. The current model, which i know this PR is mostly just moving code around and not new code you've created, doesn't provide us any guarantees that the hardware is actually power off. Do we want to address this kind of guarantee in this PR?

Reply:
You're right, I agree that this might be a good opportunity to introduce a mechanism to verify that the hardware is indeed powered off before tests commence. 

I believe we could add some retries on failure to address this. But even then, if the retries fail in that case we could raise some kind of alert that requires manual intervention. We currently don't fail the test in the cleanup stage, which I think is the most accurate way to guarantee and raise alarm to machines not powered off. I'd love to hear your thoughts on that and if that should change.

That said, I would like to keep this PR focused on the current task and maybe address the guarantee issue in another. Let me know if you think this works for you.